File: 115-distinct-subsequences/distinct-subsequences.py
class Solution:
    def numDistinct(self, s: str, t: str) -> int:
        n = len(s) # x
        m = len(t) # y
        dp = [[0] * (m+1) for _ in range(n+1)]

        # initialize
        for i in range(0, n+1):
            dp[i][0] = 1
        for j in range(1, m+1):
            dp[0][j] = 0

        for x in range(1,n + 1):
            for y in range(1, m + 1):
                if s[x-1] == t[y-1]:
                    dp[x][y] += dp[x-1][y-1]
                dp[x][y] += dp[x-1][y]
        
        return dp[n][m]

File: 115-distinct-subsequences/distinct-subsequences.cpp
class Solution {
public:
    int numDistinct(string s, string t) {
        using u128 = __uint32_t;
        int n = s.size();
        int m = t.size();
        vector<vector<u128>> dp(n+1, vector<u128>(m+1, 0));
        for(int i = 0; i <= n;i++){
            dp[i][0] = 1;
        }

        for(int i = 1; i <= n; i++){
            for(int j = 1; j <= m; j++){
                if(s[i-1] == t[j-1]) dp[i][j] += dp[i-1][j-1];
                dp[i][j] += dp[i-1][j];
            }
        }
        //print_dp(dp);

        return dp[n][m];
    }
};